### PR TITLE
Update documentation 

### LINKED ISSUE #232: Documentation: Include README.md
**Is your feature request related to a problem? Please describe.**
Currently documentation contains text copy-pasted from `README.md`.
This is problematic for 2 reasons:
1. We may add something in one place and forget to add in another (e.g., currently **Citation** is included only in GitHub README and not mentioned in the docs).
2.  Tables become out-of-sync and we have to update them in 2 places. 

**Describe the solution you'd like**
Import markdown files directly into documentations using `m2r2`.

### PR DESCRIPTION
**Pull request by zakajd**
Closes #232 

Hi guys.
I completely reworked our documentation and managed to include current README and CONTRIBUTING files into it.
One major update is that I **changed file format** from `Markdown` to `reStructuredText`. If you are not familiar with it,
here is a[ quick tour](https://docutils.sourceforge.io/docs/user/rst/quickref.html#hyperlink-targets).

Reasons to do this:
- It's much more consistent (1 official spec vs. 30+ spec. for different flavours of Markdown)
- It has richer formatting out of the box (simple tables, TOC) and native Sphinx support without extensions
- GitHub renders it's same way as `.md` files, so no change from user experience.

Here is other relevant articles 
- https://www.zverovich.net/2016/06/16/rst-vs-markdown.html
- https://www.ericholscher.com/blog/2016/mar/15/dont-use-markdown-for-technical-docs/

## Proposed Changes
  - Re-write and re-structure `README` and `CONTRIBUTING` files
  - Re-write Sphinx build files
  - Added PIQ logo at the top left instead of text (see screenshot). There is also an option to make it transparent.
  - Added piq logo files 
  - Added fancy interactive search extension used at [readthedocs](https://docs.readthedocs.io/en/stable/)
  - Some other small fixes

## What didn't change
- `benchmark.csv`. We have lot's of open PRs with metrics, will prettify table after last one is merged
-  Functions docstrings and other fixes. There is still lot's of work to be done.  Will do it as a separate PR

doc_demo

### CODE REVIEW

File: docs/source/_static/benchmark.csv [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+VSI,0.8949 / 0.7159,0.8965 / 0.7183,0.8780 / 0.6899,0.861 / 0.678
+Content,0.7049 / 0.5173,- / -,0.7237 / 0.5326,- / -
+Style,0.5384 / 0.3720,- / -,0.6470 / 0.4646,- / -
+HaarPSI,0.8732 / 0.6923,0.8732 / 0.6923,0.8849 / 0.6995,0.885 / 0.699
+MDSI,0.8899 / 0.7123,0.8899 / 0.7123,0.8853 / 0.7023,0.885 / 0.702
+MS-GMSD,0.8121 / 0.6455,0.8139 / 0.6467,0.8523 / 0.6692,- / -
+MS-GMSDc,0.8875 / 0.7105,0.687 / 0.496,0.8697 / 0.6831,- / -
+LPIPS-VGG,0.6696 / 0.4970,0.670 / 0.497,0.7201 / 0.5313,- / -
+PieAPP,0.8355 / 0.6495,0.875 / 0.710,0.8655 / 0.6758,- / -
+DISTS,0.7077 / 0.5212,0.830 / 0.639,0.8137 / 0.6254,- / -
```
Comment (denproc) [inline docs/source/_static/benchmark.csv:23]: Fix the "-" adding "\\" or copy-paste previous table.
Screenshot 2021-02-14 at 20 00 44

File: docs/source/conf.py [hunk 2]
```diff
@@ -51,6 +52,10 @@
 # a list of builtin themes.
 #
 html_theme = 'sphinx_rtd_theme'
+html_logo = '_static/piq_logo.png'
```
Comment (denproc) [inline docs/source/conf.py:55]: Option 1 (like it is done by pytorch):
1. Let's use logo with transparent background without "by Photosynthesis Team"
2. Let's use the light-grey or white background of the readthedocs theme.

Option 2:
Use just text

I tried to use different combinations of background/logo, unfortunately none of them were satisfying enough to use. Therefore, I suggest to keep just title of the library.

File: docs/source/readme.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+.. include:: ../../README.rst
```
Comment (denproc) [inline docs/source/readme.rst:1]:
```suggestion
PyTorch Image Quality (PIQ) Documentation
=========================================
.. include:: ../../README.rst
```

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+List of metrics
+---------------
+
+Full Reference
+^^^^^^^^^^^^^^
+
+===========  ======  ==========
+Acronym      Year    Metric
+===========  ======  ==========
+PSNR         -       `Peak Signal-to-Noise Ratio <https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio>`_
```
Comment (denproc) [inline README.rst:157]: Add "\\" before "-" here and for further instances.
```suggestion
PSNR         \-       `Peak Signal-to-Noise Ratio <https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio>`_
```

File: README.rst [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+PyTorch Image Quality
+=====================
+
+.. image:: https://user-images.githubusercontent.com/22414094/106116361-df8a3c80-6162-11eb-83df-f5838110e614.png
+
+..
+
+  PyTorch Image Quality (PIQ) is not endorsed by Facebook, Inc.;
+
+  PyTorch, the PyTorch logo and any related marks are trademarks of Facebook, Inc.
```
Comment (denproc) [inline README.rst:11]: The readthedocs does not show it as a note.

File: README.rst [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+* Unified interface, which is easy to use and extend.
+* Written on pure PyTorch with bare minima of additional dependencies.
+* Extensive user input validation. You code will not crash in the middle of the training.
+* Fast (GPU computations available) and reliable.
+* Most metrics can be backpropagated for model optimization.
+* Supports python 3.6-3.8.
+
+PIQ was initially named `PhotoSynthesis.Metrics <https://pypi.org/project/photosynthesis-metrics/0.4.0/>`_
+
+.. contents:: Table of Contents
```
Comment (denproc) [inline README.rst:49]: I suggest to use heading for table go content. It is lost in plain text.
image

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+   python3 tests/results_benchmark.py --dataset tid2013 --metrics SSIM MS-SSIM --path ~/datasets/tid2013 --batch_size 16
+
+We report `Spearman's Rank Correlation cCoefficient <https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient>`_ (SRCC) 
+and `Kendall rank correlation coefficient <https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient>`_ (KRCC). 
+We do not report `Pearson linear correlation coefficient <https://en.wikipedia.org/wiki/Pearson_correlation_coefficient>`_ (PLCC) 
+as it's highly dependent on fitting method and is biased towards simple examples.
+
+For metrics that can take greyscale or colour images, ``c`` means chromatic version.
+
+.. csv-table:: Benchmark results
```
Comment (denproc) [inline README.rst:220]: The default style of the table title looks sad. Let's use separate heading or no heading at all.

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,260 @@
+=====================
+PyTorch Image Quality
+=====================
+
+.. image:: https://user-images.githubusercontent.com/22414094/106116361-df8a3c80-6162-11eb-83df-f5838110e614.png
```
Comment (denproc) [inline README.rst:5]: To unify the proposed approach, the local copy of the image from docs/_static folder should be used.

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+We report `Spearman's Rank Correlation cCoefficient <https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient>`_ (SRCC) 
+and `Kendall rank correlation coefficient <https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient>`_ (KRCC). 
+We do not report `Pearson linear correlation coefficient <https://en.wikipedia.org/wiki/Pearson_correlation_coefficient>`_ (PLCC) 
+as it's highly dependent on fitting method and is biased towards simple examples.
+
+For metrics that can take greyscale or colour images, ``c`` means chromatic version.
+
+.. csv-table:: Benchmark results
+   :file: docs/source/_static/benchmark.csv
```
Comment (denproc) [inline README.rst:221]: Such reference works for the documentation, but does not for GitHub README.

File: docs/source/index.rst [hunk 10]
```diff
@@ -1,70 +1,29 @@
-.. PyTorch Image Quality (PIQ) documentation master file, created by
-   sphinx-quickstart on Tue Jan 26 18:15:25 2021.
-   You can adapt this file completely to your liking, but it should at least
-   contain the root `toctree` directive.
+.. include:: ../../README.rst
```
Comment (denproc) [inline docs/source/index.rst:1]:
```suggestion
.. include:: readme.rst
```

File: README.rst [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+For a full list of examples, see `feature metrics <https://github.com/photosynthesis-team/piq/blob/master/examples/feature_metrics.py>`_ examples.
+
+
+List of metrics
+---------------
+
+Full Reference
+^^^^^^^^^^^^^^
+
+===========  ======  ==========
```
Comment (denproc) [inline README.rst:154]: The tables with list of metrics should be imported as csv files similarly to benchmark to simplify the reuse in different places.

File: docs/source/index.rst [hunk 12]
```diff
@@ -1,70 +1,29 @@
...
-
-   PyTorch, the PyTorch logo and any related marks are trademarks of Facebook, Inc.
+.. toctree::
+   :maxdepth: 2
+   :hidden:
+   :caption: Documentation
 
-`PyTorch Image Quality (PIQ) <https://github.com/photosynthesis-team/piq>`_ is a collection of measures and metrics for image quality assessment.
-PIQ allows integrating most popular image quality metric for you ``torch`` project.
+   readme.rst
```
Comment (denproc) [inline docs/source/index.rst:8]:
```suggestion
   readme.rst
   contributing.rst
```

File: docs/source/index.rst [hunk 13]
```diff
@@ -1,70 +1,29 @@
...
 
 .. toctree::
    :maxdepth: 2
    :hidden:
```
Comment (denproc) [inline docs/source/index.rst:12]:
```suggestion
```

File: docs/source/index.rst [hunk 14]
```diff
@@ -1,70 +1,29 @@
...
-   :caption: GETTING STARTED
 
-   overview.rst
-   installation.rst
-   examples.rst
    contributing.rst
```
Comment (denproc) [inline docs/source/index.rst:14]:
```suggestion
```

**Review by denproc — CHANGES_REQUESTED** [review]
The documentation is hosted [here](https://piq.readthedocs.io/en/docs-include_markdown/index.html).

## General discussion:
I like to keep the single copy of the text used for GitHub and readthedocs both.
However, I disagree with an approach to reuse the readme/contribution as is. 
    - The project overview as a documentation becomes quite heavy with a lot of info on a single page. See [Jupyter documentation](https://jupyterlab.readthedocs.io/en/latest/index.html) as an example. 
    - The links used in GitHub versions and docs versions are different. The links from GitHub based files should lead to GitHub, while documentation ones should keep used at readthdocs. For example, link to contribution guide.
    - There is no need in “Contribution” in documentation as there is “Contribution guide”. However, this section is needed in GitHub based Readme.
    - Badges are not needed in the documentation.
    - The GitHub README does not show the benchmark table due to wrong relative path.


Possible solution:
Create universal sections of smaller size, which can be reused in README and documentation differently. Set up different cross-reference links for GitHub and readthedocs.

File: docs/source/conf.py [hunk 15]
```diff
@@ -51,6 +52,10 @@
...
+html_theme_options = {
+    'logo_only': True,
+}
```
Comment (denproc) [inline docs/source/conf.py:58]: Option 1 - light-grey background 
```suggestion
html_theme_options = {
    'logo_only': True,
    'style_nav_header_background': '#FAFAFA',
}
```

**Review by snk4tr — CHANGES_REQUESTED** [review]
Hi @zakajd. Thanks for the refactoring. I like the idea to make the text universal. However, I see some disadvantages of the approach you chose, most of which are already described by @denproc. 
I would like to additionally emphasise that the table with benchmarks in the README is a great thing and should not be removed for the sake of transferability.

**Comment by zakajd** [comment]
@denproc  @snk4tr 
I tried to apply all your suggestions, but there are 2 problems
- GitHub doesn't allow `include` or `table` command for a [security reasons](https://github.com/github/markup/issues/172). So it's impossible to split everything into small chunks and import as needed.
- Issues with incorrect links is rooted somewhere in sphinx build process so it's not so easy to change them. 

Here is what was done:
1. Applied one hacky way to include parts of README into docs as described [here](https://muffinresearch.co.uk/selectively-including-parts-readme-rst-in-your-docs/)
2. Moved table from csv into README
3. Re-structered index and overview files to include only necessary parts (as discussed with @denproc)

File: README.rst [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+MS-GMSD      0.8121 / 0.6455    0.8139 / 0.6467 `MS-GMSD`_        0.8523 / 0.6692    \- / -
+MS-GMSDc     0.8875 / 0.7105    0.687 / 0.496 `MS-GMSD`_          0.8697 / 0.6831    \- / -
+LPIPS-VGG    0.6696 / 0.4970    0.670 / 0.497  `DISTS`_           0.7201 / 0.5313    \- / - 
+PieAPP       0.8355 / 0.6495    0.875 / 0.710 `DISTS`_            0.8655 / 0.6758    \- / -
+DISTS        0.7077 / 0.5212    0.830 / 0.639 `DISTS`_            0.8137 / 0.6254    \- / - 
+===========  =================  ================================  =================  ================================
+
+.. _TID2013: http://www.ponomarenko.info/tid2013.htm
+.. _KADID10k: http://database.mmsp-kn.de/kadid-10k-database.html
+.. _link: eval2019 survey
```
Comment (denproc) [inline README.rst:267]: Insert link here

**Review by denproc — CHANGES_REQUESTED** [review]
@zakajd great rework, nice catch with including files between particular markers. 
I added high-res logo and structure of the index that I meant during discussion. Feel free to change it in case of any inconveniences.

I assume the only issue is missed link for some metric benchmarks.

File: README.rst (hunk 16 above)
Comment (zakajd) [reply README.rst]: Done 👍

**Comment by zakajd** [comment]
There is also 1 code smell due to one commented line in conf.py. Please fix this before merge

**Review by denproc — APPROVED** [review]

**Review by denproc — APPROVED** [review]

**Pull request merged**